psi4/src/psi4/libfock/COSK.cc

```diff
@@ -185,85 +185,87 @@ COSK::COSK(std::shared_ptr<BasisSet> primary, Options& options) : SplitJK(primar
-    // Original Nesse COSX formulation does not support negative grid weights
-    // which can happen with certain grid configurations
-    // the Psi4 COSX implementation is slightly modified to work with negative grid weights
-    // See https://github.com/psi4/psi4/issues/2890 for discussion
-    auto warning_printed_init = false;
-    for (const auto &init_block : grid_init_->blocks()) {
-        const auto w = init_block->w();
-        for (int ipoint = 0; ipoint < init_block->npoints(); ++ipoint) {
-            if (w[ipoint] < 0.0) {
-                outfile->Printf("  INFO: The definition of the current initial grid includes negative weights, which the original COSX formulation does not support!\n    If this is of concern, please choose another initial grid through adjusting either COSX_PRUNING_SCHEME or COSX_SPHERICAL_POINTS_INITIAL.\n\n");
-                warning_printed_init = true;
-                break;
+    // for now, we use two COSX grids:
+    //   - a small DFTGrid for the pre-converged SCF iterations
+    //   - a large DFTGrid for the final SCF iteration
+    grids_["Initial"] = nullptr;
+    grids_["Final"] = nullptr;
+
+    for (auto& [ gridname, grid ] : grids_) {
+        std::string gridname_uppercase = gridname;
+        std::transform(gridname.begin(), gridname.end(), gridname_uppercase.begin(), ::toupper);
+
+        std::string gridname_lowercase = gridname;
+        std::transform(gridname.begin(), gridname.end(), gridname_lowercase.begin(), ::tolower);
+
+        // initialize grid
+        // TODO: specify bool "DFT_REMOVE_DISTANT_POINTS" in the DFTGrid constructors
+        std::map<std::string, std::string> grid_str_options = {
+            {"DFT_PRUNING_SCHEME", options_.get_str("COSX_PRUNING_SCHEME")},
+            {"DFT_RADIAL_SCHEME",  "TREUTLER"},
+            {"DFT_NUCLEAR_SCHEME", "TREUTLER"},
+            {"DFT_GRID_NAME",      ""},
+            {"DFT_BLOCK_SCHEME",   "OCTREE"},
+        };
+
+        std::map<std::string, int> grid_int_options = {
+            {"DFT_SPHERICAL_POINTS", options_.get_int("COSX_SPHERICAL_POINTS_" + gridname_uppercase)},
+            {"DFT_RADIAL_POINTS",    options_.get_int("COSX_RADIAL_POINTS_" + gridname_uppercase)},
+            {"DFT_BLOCK_MIN_POINTS", 100},
+            {"DFT_BLOCK_MAX_POINTS", 256},
+        };
+
+        std::map<std::string, double> grid_float_options = {
+            {"DFT_BASIS_TOLERANCE",   options_.get_double("COSX_BASIS_TOLERANCE")},
+            {"DFT_BS_RADIUS_ALPHA",   1.0},
+            {"DFT_PRUNING_ALPHA",     1.0},
+            {"DFT_BLOCK_MAX_RADIUS",  3.0},
+            {"DFT_WEIGHTS_TOLERANCE", 1e-15},
+        };
+        grid = std::make_shared<DFTGrid>(primary_->molecule(), primary_, grid_int_options, grid_str_options, grid_float_options, options_);
+
+        // Print out warning if grid with negative grid weights is used
+        // Original Nesse COSX formulation does not support negative grid weights
```

Comment:
Neese

Reply:
Ah, my bad! I will fix this!

Reply:
I removed the associated warning entirely, so this issue is fixed by proxy.